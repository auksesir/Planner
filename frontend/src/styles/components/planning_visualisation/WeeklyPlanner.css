/* WeeklyPlanner.css */

.weekly-planner {
  text-align: center;
  width: 100%;
  overflow-x: visible;
  padding-bottom: 80px;
  min-height: 100vh;
  box-sizing: border-box;
}

h1 {
  font-size: 2rem;
  margin-bottom: 20px;
  color: #f0e6dd89;
}

.date-selectors {
  margin-bottom: 20px;
}

.date-selectors label {
  color: #f0e6ddd4;
  margin: 0 10px;
}

.month-select,
.year-select {
  padding: 10px;
  font-size: 1rem;
  background-color: #f0cbaf1a;
  color: #f0e6dd;
  border: 1px solid #ddd;
  border-radius: 0.25rem;
  cursor: pointer;
  transition: background-color 0.3s ease;
}

.month-select:hover,
.year-select:hover {
  background-color: #f1f1f100;
}

.weekly-nav {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 0 1rem;
}

.weekly-nav button {
  background-color: #f0cbaf1a;
  color: #ffffff68;
  border: 1px solid rgba(255, 255, 255, 0.3);
  padding: 0.5rem 0.9rem;
  font-size: 1rem;
  cursor: pointer;
  border-radius: 0.6rem;
  transition: all 0.3s ease;
}

.weekly-nav button:hover {
  background-color: #1542f7;
  color: rgba(255, 255, 255, 0.879);
}

.view-toggle {
  background-color: #1542f718;
}

.week-grid {
  display: grid;
  grid-template-columns: repeat(7, minmax(170px, 1fr));
  gap: 1px;
  height: auto;
  min-height: calc(100vh - 350px);
  margin-bottom: 40px;
}

.current-week-column {
  background-color: rgba(59, 131, 246, 0.151);
  box-shadow: 0 0 0.3rem 0.1rem rgba(59, 131, 246, 0.326);
}

.week-days-strip {
  display: grid;
  grid-template-columns: repeat(7, 1fr);
  gap: 1px;
  margin-bottom: 1rem;
}

.day-selector {
  padding: 10px;
  cursor: pointer;
  transition: all 0.3s ease;
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 0.2rem;
  color: #f0e6ddd4;
}

.day-selector:hover {
  background-color: rgba(255, 255, 255, 0.05);
}

.day-selector.selected {
  color: #fff;
  background-color: rgba(21, 66, 247, 0.1);
}

.daily-planner-container {
  flex-grow: 1;
  overflow-y: auto;
  padding-bottom: 90px;
  width: min(70%, 20rem);
  margin-inline: auto;
  margin-top: 60px;
  position: relative;
  top: 0;
}

.weekly-scroll-container {
  overflow-x: scroll;
  overflow-y: auto;
  width: 100%;
  max-width: 1200px;
  margin: 0 auto;
  height: calc(100vh - 250px);
  padding-bottom: 1rem;
  margin-bottom: 60px;
  position: relative;
  box-sizing: border-box;
  -webkit-overflow-scrolling: touch;
  scrollbar-width: auto;
}

.weekly-content-wrapper {
  width: 1680px;
  position: relative;
  padding-bottom: 20px;
}

.weekly-scroll-container::-webkit-scrollbar,
.week-grid::-webkit-scrollbar,
.daily-planner-container::-webkit-scrollbar {
  width: 8px;
  height: 12px;
}

.weekly-scroll-container::-webkit-scrollbar-track,
.week-grid::-webkit-scrollbar-track,
.daily-planner-container::-webkit-scrollbar-track {
  background: rgba(255, 255, 255, 0.05);
  border-radius: 4px;
}

.weekly-scroll-container::-webkit-scrollbar-thumb,
.week-grid::-webkit-scrollbar-thumb,
.daily-planner-container::-webkit-scrollbar-thumb {
  background: rgba(255, 255, 255, 0.2);
  border-radius: 4px;
  border: 1px solid rgba(255, 255, 255, 0.1);
}

.weekly-scroll-container::-webkit-scrollbar-thumb:hover,
.week-grid::-webkit-scrollbar-thumb:hover,
.daily-planner-container::-webkit-scrollbar-thumb:hover {
  background: rgba(255, 255, 255, 0.3);
}

.calendar-grid {
  display: grid;
  grid-template-columns: repeat(7, 1fr);
  gap: 4px;
  margin-bottom: 1rem;
  width: 100%;
  max-width: 100%;
  overflow-x: hidden;
  overflow-y: hidden;
  position: relative;
}

@media (max-width: 1200px) {
  .week-grid {
    grid-template-columns: repeat(7, 200px);
  }
  
  .weekly-scroll-container {
    height: calc(100vh - 270px);
    margin-bottom: 70px;
  }
  
  .weekly-planner {
    padding-bottom: 90px;
  }
}

@media (max-width: 768px) {
  .weekly-content-wrapper {
    min-width: 1400px;
  }
  
  .daily-planner-container {
    width: 90%;
    padding-bottom: 100px;
  }
  
  .date-selectors {
    flex-direction: column;
    gap: 0.5rem;
  }
  
  .weekly-scroll-container {
    height: calc(100vh - 290px);
    margin-bottom: 80px;
    padding-bottom: 1.5rem;
  }
  
  .weekly-planner {
    padding-bottom: 100px;
  }
  
  .week-grid {
    margin-bottom: 60px;
    min-height: calc(100vh - 380px);
  }
}

@media (max-width: 640px) {
  .weekly-planner::after {
    content: '';
    position: absolute;
    right: 0;
    top: 200px;
    width: 30px;
    height: calc(100vh - 300px);
    background: linear-gradient(to right, transparent, rgba(8, 5, 39, 0.7));
    pointer-events: none;
    z-index: 10;
  }
  
  .weekly-scroll-container {
    height: calc(100vh - 310px);
    margin-bottom: 90px;
  }
  
  .weekly-planner {
    padding-bottom: 110px;
  }
  
  .week-grid {
    margin-bottom: 80px;
  }
}