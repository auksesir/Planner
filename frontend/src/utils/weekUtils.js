import { format } from 'date-fns';
import { getRemindersForWeek, getTasksForWeek } from '../api/api';

// Process and organize tasks by day with logging
export const organizeTasksByDay = (tasks) => {
  
  const organized = tasks.reduce((acc, task) => {
    const day = format(new Date(task.selectedDay), 'yyyy-MM-dd');
    
    if (!acc[day]) {
      acc[day] = [];
    }
    acc[day].push(task);
    return acc;
  }, {});

  return organized;
};

// In weekUtils.js - Fix the organizeRemindersByDay function
export const organizeRemindersByDay = (reminders) => {
  
  const organized = reminders.reduce((acc, reminder) => {
    // Log the incoming reminder

    try {
      // Use the selectedDay property for organizing by day
      const reminderDate = new Date(reminder.selectedDay);
      if (isNaN(reminderDate.getTime())) {
        return acc;
      }
      
      const day = format(reminderDate, 'yyyy-MM-dd');

      if (!acc[day]) {
        acc[day] = [];
      }
      
      // Make sure we have valid times
      const reminderTime = reminder.selectedTime || reminder.time;
      if (!reminderTime) {
        return acc;
      }

      acc[day].push({
        ...reminder,
        selectedTime: reminderTime // Ensure selectedTime is always set
      });
    } catch (error) {
      ('Error processing reminder:', error, reminder);
    }

    return acc;
  }, {});

  // Log the final organized structure
  return organized;
};

// Fetch and process week data with enhanced error handling and logging
export const fetchWeekData = async (weekDays) => {
  try {    
    const startDate = format(weekDays[0], 'yyyy-MM-dd');
    const endDate = format(weekDays[6], 'yyyy-MM-dd');
    
    const [weekTasks, weekReminders] = await Promise.all([
      getTasksForWeek(startDate, endDate),
      getRemindersForWeek(startDate, endDate)
    ]);

    // Organize the data
    const organizedTasks = organizeTasksByDay(weekTasks);
    const organizedReminders = organizeRemindersByDay(weekReminders);

    return {
      tasks: organizedTasks,
      reminders: organizedReminders
    };
  } catch (error) {
    ('Error in fetchWeekData:', error);
    // Return empty objects instead of throwing to prevent UI crashes
    return {
      tasks: {},
      reminders: {}
    };
  }
};