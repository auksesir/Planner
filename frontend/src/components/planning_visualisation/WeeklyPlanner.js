import { addDays, format, getHours, isSameDay, startOfWeek } from 'date-fns';
import { useEffect, useMemo, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { setSelectedDayFlag } from '../../redux/tasks/actions/dailyPlannerFlagActions';
import { setWeeklyGridOpen } from '../../redux/tasks/actions/weeklyGridActions';
import '../../styles/components/planning_visualisation/WeeklyPlanner.css';
import { getMonthsArray, getYearsArray, groupItemsByTimeSlot } from '../../utils/timeUtils';
import { fetchWeekData } from '../../utils/weekUtils';
import DeleteTaskModal from '../planning_utilities/DeleteTaskModal';
import Sphere from '../planning_utilities/Sphere';
import DailyPlanner from './DailyPlanner';
import DayCell from './DayCell';

/**
 * WeeklyPlanner Component
 * 
 * A comprehensive weekly planning interface that provides both grid and individual day views.
 * Features dynamic height calculations, event management, and seamless integration with daily planning.
 * 
 * Key Features:
 * - Dual view modes: Weekly grid view and single day detailed view
 * - Dynamic hour height calculations based on event density
 * - Real-time data synchronization with Redux state management
 * - Event CRUD operations with support for recurring items
 * - Responsive design with horizontal scrolling for weekly view
 * - Date navigation with year/month selectors and week navigation
 * - Integration with task and reminder management systems
 * 
 * @param {Function} setTaskToEdit - Callback to set a task for editing
 * @param {Function} setReminderToEdit - Callback to set a reminder for editing
 */
const WeeklyPlanner = ({ setTaskToEdit, setReminderToEdit }) => {
  
  // ==================== CORE STATE ====================
  
  // Date and navigation state
  const today = new Date();
  const [selectedDate, setSelectedDate] = useState(today);        // Currently selected week anchor date
  const [weekDays, setWeekDays] = useState([]);                   // Array of 7 days in current week
  const [activeDay, setActiveDay] = useState(today);              // Currently focused day in single day view
  
  // View state
  const [showAllDays, setShowAllDays] = useState(true);           // Toggle between grid view and single day view
  
  // Data state
  const [weeklyTasks, setWeeklyTasks] = useState({});             // Tasks organized by date string (YYYY-MM-DD)
  const [weeklyReminders, setWeeklyReminders] = useState({});     // Reminders organized by date string
  const [hourHeights, setHourHeights] = useState({});             // Dynamic heights for each hour based on event density
  
  // Modal and deletion state
  const [isDeleteModalOpen, setIsDeleteModalOpen] = useState(false);     // Delete confirmation modal visibility
  const [selectedItemId, setSelectedItemId] = useState(null);            // ID of item to delete
  const [selectedItem, setSelectedItem] = useState(null);                // Full object of item to delete
  const [itemType, setItemType] = useState(null);                        // 'task' or 'reminder'
  const [isRepeatingItem, setIsRepeatingItem] = useState(false);         // Whether item has repeat settings

  // ==================== REDUX INTEGRATION ====================
  
  const dispatch = useDispatch();
  const settings = useSelector(state => state.settings || {});           // User preferences for time display
  const dailyPlannerFlag = useSelector(state => state.dailyPlannerFlag); // Flag to trigger data refresh

  // ==================== COMPONENT LIFECYCLE ====================
  
  /**
   * Component Mount/Unmount Effect
   * Notifies Redux that the weekly grid is open/closed for proper state management
   */
  useEffect(() => {
    dispatch(setWeeklyGridOpen(true));
    return () => dispatch(setWeeklyGridOpen(false));
  }, [dispatch]);

  /**
   * Week Days Initialization
   * Calculates the 7 days of the current week based on selected date
   * Week starts on Sunday (weekStartsOn: 0)
   */
  useEffect(() => {
    const start = startOfWeek(selectedDate, { weekStartsOn: 0 });
    const days = [...Array(7)].map((_, i) => addDays(start, i));
    setWeekDays(days);
  }, [selectedDate]);

  // ==================== DATA REFRESH MANAGEMENT ====================
  
  /**
   * Redux Flag-Triggered Data Refresh
   * Monitors dailyPlannerFlag and refreshes weekly data when external changes occur
   * This ensures the weekly view stays synchronized with other parts of the application
   */
  useEffect(() => {
    const refreshData = async () => {
      if (weekDays.length > 0) {
        try {
          const { tasks, reminders } = await fetchWeekData(weekDays);
          setWeeklyTasks(tasks);
          setWeeklyReminders(reminders);
        } catch (error) {
          console.error('Error refreshing weekly data:', error);
        }
      }
    };
  
    // Call this function when dailyPlannerFlag changes to true
    if (dailyPlannerFlag === true) {
      refreshData();
      // Reset the flag after initiating the refresh
      dispatch(setSelectedDayFlag(false));
    }
  }, [dailyPlannerFlag, weekDays, dispatch]);

  // ==================== DYNAMIC HEIGHT CALCULATIONS ====================
  
  /**
   * Calculate Hour Heights Based on Event Density
   * 
   * Dynamically calculates the height for each hour slot based on the maximum number
   * of concurrent events across all days of the week. This ensures proper spacing
   * and prevents overlapping in the weekly grid view.
   * 
   * Algorithm:
   * 1. For each hour (0-23), check all 7 days
   * 2. Count tasks and reminders that occur during that hour
   * 3. Find the maximum event count across all days for that hour
   * 4. Calculate height: base 60px + 24px per additional event
   */
  const calculateHourHeights = useMemo(() => {
    if (Object.keys(weeklyTasks).length === 0) return {};
    
    const heights = {};
    
    // Get all visible hours from settings (currently all 24 hours)
    const visibleHours = Array.from({ length: 24 }, (_, i) => i).filter(hour => {
      // Future enhancement: filter based on user settings
      return true; // For now, include all hours
    });
    
    visibleHours.forEach(hour => {
      let maxEventsCount = 0;
      
      // Check each day for this hour
      weekDays.forEach(day => {
        const dayStr = format(day, 'yyyy-MM-dd');
        const dayTasks = weeklyTasks[dayStr] || [];
        const dayReminders = weeklyReminders[dayStr] || [];
        
        let eventsInHour = 0;
        
        // Count tasks that span or occur during this hour
        dayTasks.forEach(task => {
          const taskStartHour = getHours(new Date(task.startTime));
          const taskEndHour = getHours(new Date(task.endTime));
          
          if (hour >= taskStartHour && hour <= taskEndHour) {
            eventsInHour++;
          }
        });
        
        // Count reminders that occur during this hour
        dayReminders.forEach(reminder => {
          const reminderTime = reminder.selectedTime || reminder.time;
          if (reminderTime) {
            const reminderHour = getHours(new Date(reminderTime));
            if (hour === reminderHour) {
              eventsInHour++;
            }
          }
        });
        
        maxEventsCount = Math.max(maxEventsCount, eventsInHour);
      });
      
      // Calculate height: base height + additional height per event
      // Base height of 60px + 24px per event after the first
      heights[hour] = maxEventsCount <= 1 ? 60 : 60 + ((maxEventsCount - 1) * 24);
    });
    
    return heights;
  }, [weeklyTasks, weeklyReminders, weekDays]);

  // ==================== INITIAL DATA LOADING ====================
  
  /**
   * Fetch Weekly Data on Week Change
   * Loads tasks and reminders for all 7 days when week changes
   */
  useEffect(() => {
    if (weekDays.length > 0) {
      const loadWeekData = async () => {
        try {
          const { tasks, reminders } = await fetchWeekData(weekDays);
          console.log("Tasks by day:", tasks);
          
          // Verify data structure is correct
          Object.keys(tasks).forEach(dayKey => {
            console.log(`Day ${dayKey} has ${tasks[dayKey]?.length || 0} tasks`);
          });
          
          setWeeklyTasks(tasks);
          setWeeklyReminders(reminders);
        } catch (error) {
          console.error('Error loading week data:', error);
        }
      };
  
      loadWeekData();
    }
  }, [weekDays]);

  /**
   * Update Hour Heights When Data Changes
   * Recalculates dynamic heights whenever tasks or reminders are updated
   */
  useEffect(() => {
    setHourHeights(calculateHourHeights);
  }, [weeklyTasks, weeklyReminders, calculateHourHeights]);

  /**
   * DOM Update and Scroll Management
   * Ensures proper scrolling behavior after state updates
   */
  useEffect(() => {    
    // Add a small delay to ensure the DOM has updated
    setTimeout(() => {
      const scrollContainer = document.getElementById('weekly-scroll');
      if (scrollContainer) {
        // Force a scroll update by slightly adjusting scroll position
        scrollContainer.scrollLeft = 0;
      }
    }, 100);
  }, [weeklyTasks, weeklyReminders, hourHeights]);

  // ==================== EVENT HANDLERS ====================
  
  /**
   * Handle day click in weekly grid view
   * Switches to single day view and sets the clicked day as active
   */
  const handleDayClick = (date) => {
    setActiveDay(date);
    setShowAllDays(false);
  };

  /**
   * Handle day selection in single day view
   * Changes the active day without changing view mode
   */
  const handleDaySelect = (date) => {
    setActiveDay(date);
  };

  /**
   * Handle event editing
   * Sets the appropriate day as active and triggers edit mode for tasks/reminders
   */
  const handleEventEdit = (event) => {
    // Update active day without changing the view mode
    const eventDate = new Date(event.selectedDay || event.startTime || event.time);
    setActiveDay(eventDate);

    if (event.type === 'task') {
      setTaskToEdit(event);
    } else if (event.type === 'reminder') {
      setReminderToEdit(event);
    }
  };

  /**
   * Handle event deletion initiation
   * Opens delete confirmation modal with event details
   */
  const handleDeleteEvent = (event) => {
    setSelectedItemId(event.id);
    setSelectedItem(event);
    setItemType(event.type);
    setIsRepeatingItem(!!event.repeatOption);
    setIsDeleteModalOpen(true);
  };

  /**
   * Close delete modal and reset related state
   */
  const closeDeleteModal = () => {
    setIsDeleteModalOpen(false);
    setSelectedItem(null);
    setSelectedItemId(null);
    setItemType(null);
    setIsRepeatingItem(false);
  };

  /**
   * Confirm and execute event deletion
   * Handles both single instance and recurring event deletion
   * 
   * @param {boolean} deleteAll - Whether to delete all instances of recurring events
   */
  const confirmDelete = async (deleteAll = false) => {
    if (selectedItemId !== null && selectedItem !== null) {
      const dateObj = selectedItem.selectedDay ? 
        new Date(selectedItem.selectedDay) : 
        new Date(selectedItem.startTime || selectedItem.time);
      
      // Format date as YYYY-MM-DD for API consistency
      const selectedDate = dateObj.toISOString().split('T')[0];
      
      try {
        // Import deletion functions dynamically to avoid circular dependencies
        const { deleteTask, deleteReminder } = await import('../../api/api');
        let result;
        
        if (itemType === 'task') {
          result = await deleteTask(selectedItemId, deleteAll, selectedDate);
        } else {
          result = await deleteReminder(selectedItemId, deleteAll, selectedDate);
        }
        
        // Set the flag to trigger a refresh across the application
        dispatch(setSelectedDayFlag(true));
      } catch (error) {
        console.error('Error deleting item:', error);
      }
    }
    
    closeDeleteModal();
  };

  // ==================== VIEW NAVIGATION ====================
  
  /**
   * Toggle between weekly grid view and single day view
   */
  const toggleView = () => {
    setShowAllDays(!showAllDays);
  };

  /**
   * Handle year change in date selector
   */
  const handleYearChange = (e) => {
    const newDate = new Date(selectedDate);
    newDate.setFullYear(parseInt(e.target.value));
    setSelectedDate(newDate);
  };

  /**
   * Handle month change in date selector
   */
  const handleMonthChange = (e) => {
    const newDate = new Date(selectedDate);
    newDate.setMonth(getMonthsArray().indexOf(e.target.value));
    setSelectedDate(newDate);
  };

  /**
   * Navigate to previous week
   */
  const handlePrevWeek = () => {
    setSelectedDate(prev => addDays(prev, -7));
  };

  /**
   * Navigate to next week
   */
  const handleNextWeek = () => {
    setSelectedDate(prev => addDays(prev, 7));
  };

  // ==================== DATA PREPARATION ====================
  
  /**
   * Prepare Active Day Data for Single Day View
   * 
   * Formats and groups tasks/reminders for the currently active day
   * Groups items by time slots for proper rendering in DailyPlanner component
   */
  const prepareActiveDayData = useMemo(() => {
    const activeDayStr = format(activeDay, 'yyyy-MM-dd');
    const activeDayTasks = weeklyTasks[activeDayStr] || [];
    const activeDayReminders = weeklyReminders[activeDayStr] || [];
    
    // Group tasks and reminders by time slot for DailyPlanner component
    const groupedTasks = groupItemsByTimeSlot(activeDayTasks, 'startTime', 'endTime');
    const groupedReminders = groupItemsByTimeSlot(activeDayReminders, 'selectedTime', 'selectedTime');
    
    return {
      dayStr: activeDayStr,
      tasks: groupedTasks,
      reminders: groupedReminders,
    };
  }, [activeDay, weeklyTasks, weeklyReminders]);

  // ==================== RENDER ====================

  return (
    <div className="weekly-planner">
      
      {/* ==================== DATE NAVIGATION ==================== */}
      <div className="date-selectors">
        <label>
          Year:
          <select
            value={format(selectedDate, 'yyyy')}
            onChange={handleYearChange}
            className="year-select"
          >
            {getYearsArray().map(year => (
              <option key={year} value={year}>{year}</option>
            ))}
          </select>
        </label>

        <label>
          Month:
          <select
            value={format(selectedDate, 'MMMM')}
            onChange={handleMonthChange}
            className="month-select"
          >
            {getMonthsArray().map(month => (
              <option key={month} value={month}>{month}</option>
            ))}
          </select>
        </label>
      </div>

      {/* Week Navigation and View Toggle */}
      <div>
        <button onClick={handlePrevWeek}>←</button>
        <button onClick={toggleView}>
          {showAllDays ? 'Show Single Day' : 'Show All Days'}
        </button>
        <button onClick={handleNextWeek}>→</button>
      </div>

      {/* ==================== CONDITIONAL VIEW RENDERING ==================== */}
      
      {showAllDays ? (
        /* Weekly Grid View - Shows all 7 days in a scrollable grid */
        <div className="weekly-scroll-container" id="weekly-scroll">
          <div className="weekly-content-wrapper">
            <div className="week-grid">
            {weekDays.map(date => {
              const dayStr = format(date, 'yyyy-MM-dd');
              return (
                <DayCell
                  key={dayStr}
                  date={date}
                  tasks={weeklyTasks[dayStr] || []}
                  reminders={weeklyReminders[dayStr] || []} 
                  onDayClick={handleDayClick}
                  settings={settings}
                  onEventEdit={handleEventEdit}
                  onEventDelete={handleDeleteEvent}
                  hourHeights={hourHeights} // Dynamic heights based on event density
                />
              );
            })}
            </div>
          </div>
        </div>
      ) : (
        /* Single Day View - Shows day selector and detailed daily planner */
        <>
          {/* Day Selector Row */}
          <div className="calendar-grid">
            {weekDays.map((date, index) => (
              <div
                key={index}
                className={`day-selector ${isSameDay(date, activeDay) ? 'selected' : ''}`}
                onClick={() => handleDaySelect(date)}
              >
                <div>{format(date, 'EEE')}</div>
                <div className="day-number">{format(date, 'd')}</div>
              </div>
            ))}
          </div>
          
          {/* Detailed Daily Planner */}
          <div className="daily-planner-container">
            <DailyPlanner
              day={prepareActiveDayData.dayStr}
              tasks={prepareActiveDayData.tasks}
              reminders={prepareActiveDayData.reminders}
              isCurrentDay={false} // This is always false since it's a weekly view
              setTaskToEdit={setTaskToEdit}
              setReminderToEdit={setReminderToEdit}
              context="weekly" // Context for styling differences
            />
          </div>
        </>
      )}
      
      {/* ==================== MODAL DIALOGS ==================== */}
      
      {/* Delete Confirmation Modal */}
      <DeleteTaskModal
        isOpen={isDeleteModalOpen}
        isRepeatingTask={isRepeatingItem}
        onClose={closeDeleteModal}
        onDeleteSingle={() => confirmDelete(false)}  // Delete single instance
        onDeleteAll={() => confirmDelete(true)}      // Delete all instances
        itemType={itemType}
      />
      
      {/* Sphere Component - Additional UI element */}
      <Sphere />
    </div>
  );
};

export default WeeklyPlanner;