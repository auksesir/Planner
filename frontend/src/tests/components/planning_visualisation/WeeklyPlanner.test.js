// @ts-nocheck
import '@testing-library/jest-dom';
import { act, fireEvent, render, screen } from '@testing-library/react';
import WeeklyPlanner from '../../../components/planning_visualisation/WeeklyPlanner';
import { fetchWeekData } from '../../../utils/weekUtils';

// Mock console to suppress warnings
jest.spyOn(console, 'error').mockImplementation(() => {});

// Mock date-fns functions
jest.mock('date-fns', () => ({
  addDays: jest.fn(date => new Date(date.getTime() + 24 * 60 * 60 * 1000)),
  format: jest.fn((date, formatStr) => {
    if (formatStr === 'yyyy-MM-dd') return '2023-07-20';
    if (formatStr === 'EEE') return 'Thu';
    if (formatStr === 'd') return '20';
    if (formatStr === 'h:mm a') return '10:00 AM';
    return date.toISOString();
  }),
  startOfWeek: jest.fn(date => {
    // Return array of dates for the week
    const result = new Date(date);
    result.setDate(result.getDate() - result.getDay()); // Set to Sunday
    return result;
  }),
  getHours: jest.fn(() => 10),
  isSameDay: jest.fn(() => true),
  getMinutes: jest.fn(() => 0)
}));

// Mock Redux
jest.mock('react-redux', () => ({
  useDispatch: () => jest.fn(),
  useSelector: jest.fn().mockImplementation(selector => {
    // Mock state for useSelector
    const mockState = {
      settings: {
        startHour: '08:00 AM',
        endHour: '06:00 PM',
        hiddenHours: []
      },
      dailyPlannerFlag: false
    };
    return selector(mockState);
  }),
  connect: () => Component => Component
}));

// Mock weekUtils
jest.mock('../../../utils/weekUtils', () => ({
  fetchWeekData: jest.fn().mockImplementation(async () => ({
    tasks: { '2023-07-20': [] },
    reminders: { '2023-07-20': [] }
  })),
  organizeTasksByDay: jest.fn(tasks => ({ '2023-07-20': tasks || [] })),
  organizeRemindersByDay: jest.fn(reminders => ({ '2023-07-20': reminders || [] }))
}));

// Mock components
jest.mock('../../../components/planning_visualisation/DailyPlanner', () => ({
  __esModule: true,
  default: function MockDailyPlanner(props) {
    return (
      <div data-testid="daily-planner">
        <div>Day: {props.day}</div>
        <div>Tasks: {Object.keys(props.tasks || {}).length}</div>
        <div>Reminders: {Object.keys(props.reminders || {}).length}</div>
        <div>Context: {props.context || 'default'}</div>
      </div>
    );
  }
}));

jest.mock('../../../components/planning_visualisation/DayCell', () => function MockDayCell(props) {
  return (
    <div 
      data-testid="day-cell"
      onClick={() => props.onDayClick && props.onDayClick(props.date)}
    >
      <div>Day: Thu</div>
      <div>Date: 20</div>
    </div>
  );
});

jest.mock('../../../components/planning_utilities/Sphere', () => function MockSphere() {
  return <div data-testid="sphere" />;
});

jest.mock('../../../components/planning_utilities/DeleteTaskModal', () => function MockDeleteTaskModal(props) {
  return props.isOpen ? <div data-testid="delete-modal">Mock Delete Modal</div> : null;
});

describe('WeeklyPlanner Component', () => {
  const mockSetTaskToEdit = jest.fn();
  const mockSetReminderToEdit = jest.fn();
  
  beforeEach(() => {
    jest.clearAllMocks();
    fetchWeekData.mockImplementation(async () => ({
      tasks: { '2023-07-20': [] },
      reminders: { '2023-07-20': [] }
    }));
  });

  test('renders without crashing', async () => {
    await act(async () => {
      render(
        <WeeklyPlanner 
          setTaskToEdit={mockSetTaskToEdit} 
          setReminderToEdit={mockSetReminderToEdit} 
        />
      );
    });
    expect(screen.getAllByTestId('day-cell').length).toBeGreaterThan(0);
  });

  test('handles day click', async () => {
    await act(async () => {
      render(
        <WeeklyPlanner 
          setTaskToEdit={mockSetTaskToEdit} 
          setReminderToEdit={mockSetReminderToEdit} 
        />
      );
    });

    const dayCells = screen.getAllByTestId('day-cell');
    await act(async () => {
      fireEvent.click(dayCells[0]);
    });
    expect(screen.queryByTestId('daily-planner')).toBeInTheDocument();
  });

  test('toggles view between all days and single day', async () => {
    await act(async () => {
      render(
        <WeeklyPlanner 
          setTaskToEdit={mockSetTaskToEdit} 
          setReminderToEdit={mockSetReminderToEdit} 
        />
      );
    });

    expect(screen.getAllByTestId('day-cell').length).toBeGreaterThan(1);
    
    const toggleButton = screen.getByText('Show Single Day');
    await act(async () => {
      fireEvent.click(toggleButton);
    });
    
    expect(screen.getByText('Show All Days')).toBeInTheDocument();
    expect(screen.queryByTestId('daily-planner')).toBeInTheDocument();
    
    await act(async () => {
      fireEvent.click(screen.getByText('Show All Days'));
    });
    
    expect(screen.getByText('Show Single Day')).toBeInTheDocument();
    expect(screen.getAllByTestId('day-cell').length).toBeGreaterThan(1);
  });
  
  test('handles empty data state', async () => {
    fetchWeekData.mockImplementation(async () => ({
      tasks: {},
      reminders: {}
    }));

    await act(async () => {
      render(
        <WeeklyPlanner 
          setTaskToEdit={mockSetTaskToEdit} 
          setReminderToEdit={mockSetReminderToEdit} 
        />
      );
    });

    expect(screen.getAllByTestId('day-cell').length).toBeGreaterThan(0);
  });

  test('matches snapshot', async () => {
    const { asFragment } = await act(async () => {
      return render(
        <WeeklyPlanner 
          setTaskToEdit={mockSetTaskToEdit} 
          setReminderToEdit={mockSetReminderToEdit} 
        />
      );
    });
    
    expect(asFragment()).toMatchSnapshot();
  });
});