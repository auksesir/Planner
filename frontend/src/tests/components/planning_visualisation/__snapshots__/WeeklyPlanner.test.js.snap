// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeeklyPlanner Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="weekly-planner"
  >
    <div
      class="date-selectors"
    >
      <label>
        Year:
        <select
          class="year-select"
        >
          <option
            value="2020"
          >
            2020
          </option>
          <option
            value="2021"
          >
            2021
          </option>
          <option
            value="2022"
          >
            2022
          </option>
          <option
            value="2023"
          >
            2023
          </option>
          <option
            value="2024"
          >
            2024
          </option>
          <option
            value="2025"
          >
            2025
          </option>
          <option
            value="2026"
          >
            2026
          </option>
          <option
            value="2027"
          >
            2027
          </option>
          <option
            value="2028"
          >
            2028
          </option>
          <option
            value="2029"
          >
            2029
          </option>
          <option
            value="2030"
          >
            2030
          </option>
        </select>
      </label>
      <label>
        Month:
        <select
          class="month-select"
        >
          <option />
          <option />
          <option />
          <option />
          <option />
          <option />
          <option />
          <option />
          <option />
          <option />
          <option />
          <option />
        </select>
      </label>
    </div>
    <div>
      <button>
        ←
      </button>
      <button>
        Show Single Day
      </button>
      <button>
        →
      </button>
    </div>
    <div
      class="weekly-scroll-container"
      id="weekly-scroll"
    >
      <div
        class="weekly-content-wrapper"
      >
        <div
          class="week-grid"
        >
          <div
            data-testid="day-cell"
          >
            <div>
              Day: Thu
            </div>
            <div>
              Date: 20
            </div>
          </div>
          <div
            data-testid="day-cell"
          >
            <div>
              Day: Thu
            </div>
            <div>
              Date: 20
            </div>
          </div>
          <div
            data-testid="day-cell"
          >
            <div>
              Day: Thu
            </div>
            <div>
              Date: 20
            </div>
          </div>
          <div
            data-testid="day-cell"
          >
            <div>
              Day: Thu
            </div>
            <div>
              Date: 20
            </div>
          </div>
          <div
            data-testid="day-cell"
          >
            <div>
              Day: Thu
            </div>
            <div>
              Date: 20
            </div>
          </div>
          <div
            data-testid="day-cell"
          >
            <div>
              Day: Thu
            </div>
            <div>
              Date: 20
            </div>
          </div>
          <div
            data-testid="day-cell"
          >
            <div>
              Day: Thu
            </div>
            <div>
              Date: 20
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      data-testid="sphere"
    />
  </div>
</DocumentFragment>
`;
